annotated.go

```
@@ -89,3 +90,263 @@ func (a Annotated) String() string {
+// equivalent to,
+//
+//   struct {
+//     fx.In
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (pt paramTagsAnnotation) apply(ann *annotations) error {
+	if ann.annotatedIn {
+		return fmt.Errorf("cannot apply more than one line of ParamTags")
+	}
+	ann.annotatedIn = true
+	fType := ann.fType
+
+	annotatedParams := []reflect.StructField{{
+		Name:      "In",
+		Type:      reflect.TypeOf(In{}),
+		Anonymous: true,
+	}}
+	for i := 0; i < fType.NumIn(); i++ {
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.In(i),
+		}
+		if i < len(pt.tags) {
+			structField.Tag = reflect.StructTag(pt.tags[i])
+		}
+		annotatedParams = append(annotatedParams, structField)
+	}
+
+	ann.Ins = []reflect.Type{reflect.StructOf(annotatedParams)}
+	return nil
+}
+
+// ParamTags is an Annotation that annotates the parameter(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional parameter.
+func ParamTags(tags ...string) Annotation {
+	return paramTagsAnnotation{tags}
+}
+
+type resultTagsAnnotation struct {
+	tags []string
+}
+
+var _ resultTagsAnnotation = resultTagsAnnotation{}
+
+// Given func(T1, T2, T3, ..., TN), this generates a type roughly
+// equivalent to,
+//
+//   struct {
+//     fx.Out
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (rt resultTagsAnnotation) apply(ann *annotations) error {
+	if ann.annotatedOut {
+		return fmt.Errorf("cannot apply more than one line of ResultTags")
+	}
+	ann.annotatedOut = true
+	fType := ann.fType
+
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	for i := 0; i < fType.NumOut(); i++ {
+		// guard against error results
+		if fType.Out(i) == _typeOfError {
+			continue
+		}
```

Oh, this seems slightly broken. We'll ignore errors completely:

```
fx.Annotate(func() (Foo, error), fx.ResultTags(`name:"foo"`))
```

will become

```
type whatever struct{
  fx.Out

  Field1 Foo `name:"foo"`
}

func() whatever
```

It will ignore the error.

What we want is,

```
func() (whatever, error)
```